src/mc-fle2-payload-iev.c

```diff
@@ -52,6 +55,230 @@ mc_FLE2IndexedEqualityEncryptedValueTokens_new (void)
+   return true;
+}
+
+bool
+mc_FLE2IndexedEncryptedValue_write (
+   _mongocrypt_crypto_t *crypto,
+   const bson_type_t original_bson_type,
+   const _mongocrypt_buffer_t *S_KeyId,
+   const _mongocrypt_buffer_t *ClientEncryptedValue,
+   mc_ServerDataEncryptionLevel1Token_t *token,
+   mc_FLE2IndexedEqualityEncryptedValueTokens *index_tokens,
+   _mongocrypt_buffer_t *buf,
+   mongocrypt_status_t *status)
+{
+   BSON_ASSERT_PARAM (crypto);
+   BSON_ASSERT_PARAM (index_tokens);
+   BSON_ASSERT_PARAM (S_KeyId);
+   BSON_ASSERT_PARAM (ClientEncryptedValue);
+   BSON_ASSERT_PARAM (token);
+   BSON_ASSERT_PARAM (index_tokens);
+   BSON_ASSERT_PARAM (buf);
+
+   if (ClientEncryptedValue->len == 0) {
+      CLIENT_ERR (
+         "mc_FLE2IndexedEncryptedValue_write iev must have an encrypted value");
+      return false;
+   }
+
+   if (S_KeyId->len == 0) {
+      CLIENT_ERR (
+         "mc_FLE2IndexedEncryptedValue_write iev SKeyId must have value");
+      return false;
+   }
+
+   _mongocrypt_buffer_t encryption_out;
+   _mongocrypt_buffer_init (&encryption_out);
+
+   if (!mc_fle2IndexedEncryptedValue_encrypt (crypto,
+                                              ClientEncryptedValue,
+                                              token,
+                                              index_tokens,
+                                              &encryption_out,
+                                              status)) {
+      return false;
+   }
+
+   uint32_t expected_plaintext_size = 0;
+   CHECK_AND_RETURN (safe_uint32_t_sum (
+      ClientEncryptedValue->len,
+      (uint32_t) (sizeof (uint64_t) * 2 + sizeof (uint32_t) * 3),
+      &expected_plaintext_size,
+      status));
+
+   uint32_t expected_cipher_size = _mongocrypt_fle2_calculate_ciphertext_len (
+      expected_plaintext_size, status);
+   uint32_t expected_buf_size = 0;
+   CHECK_AND_RETURN (safe_uint32_t_sum (expected_cipher_size,
+                                        (uint32_t) (1 + sizeof (S_KeyId)),
+                                        &expected_buf_size,
+                                        status));
+
+   if (buf->len < expected_buf_size) {
+      CLIENT_ERR (
+         "mc_FLE2IndexedEncryptedValue_write buf is not large enough for iev");
+      return false;
+   }
+
+   mc_writer_t writer;
+   mc_writer_init_from_buffer (&writer, buf, __FUNCTION__);
+
+   const uint8_t subtype =
+      (uint8_t) MC_SUBTYPE_FLE2IndexedEqualityEncryptedValue;
+
+   if ((original_bson_type < 0) || (original_bson_type > 0xFF)) {
+      CLIENT_ERR ("Field 't' must be a valid BSON type, got: %d",
+                  original_bson_type);
+      return false;
+   }
+
+   const uint8_t bson_type = (uint8_t) original_bson_type;
+
+   mc_writer_write_u8 (&writer, subtype, status);
+   mc_writer_write_buffer (&writer, S_KeyId, S_KeyId->len, status);
+   mc_writer_write_u8 (&writer, bson_type, status);
+
+   mc_writer_write_buffer (
+      &writer, &encryption_out, encryption_out.len, status);
```

Comment:
Missing error handling on potential failure of these functions. If guaranteed to not fail, suggest adding a comment or `BSON_ASSERT()`.